# Exploring eBay Car Sales Data

# Introduction

In this project, we will be working with a dataset of used cars from the classifieds section of the German eBay website. The data already has 50,000 sampled data points from the full dataset, originally uploaded to Kaggle. The aim of this project is to clean the data and analyze the included car listings.

In [343]:
#Import the NumPy and Pandas libraries
import numpy as np
import pandas as pd
#read in CSV file into Pandas
autos = pd.read_csv('autos.csv', encoding = 'Latin-1')

In [344]:
#observe first few and last few lines of the read file
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

#explore information about the dataset
autos.info()
autos.head()

We can see that column 'vehicletype', 'gearbox', 'model', 'fueltype', and 'notrepaireddamage' contain null values. Overall, 13 columns are listed as objects (strings), 6 are listed as integers, and there is one boolean column 'nrofpictures'. However, a boolean datatype does not correspond with the description of the column: the number of pictures in the ad. Additionally, the column names are in camelcase instead of snakecase (Python's preferred naming convention)

In [345]:
#we will now convert the column names from camelcase to snakecase and rename some of the columns to be more descriptive

autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [346]:
new_columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'ab_test',
       'vehicle_type', 'registration_year', 'gearbox', 'powerps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen']

In [347]:
autos.columns = new_columns
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price  ab_test vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  powerps   model   odometer  registration_month fuel_type  \
0    manuell      158  andere  150,000km                   3       lpg   
1  automatik      286     7er  150,000km                   6    benzin   
2    manuell      102    golf   70,000km                   7    benzin   
3  automatik       71  fortwo   70,000km                   6    benzin   
4    manuell        0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

We renamed some of the column names to be more descriptive and aligned with the information in the data dictionary:

1) yearofregistration to registrationyear

2) monthofregistration to registration_month

3) notrepaireddamage to unrepaired_damage

4) datecreated to ad_created

The remaining column names were changed from camelcase to snakecase.

In [348]:
#We will now conduct an initial data exploration to determine what other cleaning tasks need to be done
autos.describe(include = 'all')

date_crawled         name  seller offer_type  price ab_test  \
count                 50000        50000   50000      50000  50000   50000   
unique                48213        38754       2          2   2357       2   
top     2016-03-25 19:57:10  Ford_Fiesta  privat    Angebot     $0    test   
freq                      3           78   49999      49999   1421   25756   
mean                    NaN          NaN     NaN        NaN    NaN     NaN   
std                     NaN          NaN     NaN        NaN    NaN     NaN   
min                     NaN          NaN     NaN        NaN    NaN     NaN   
25%                     NaN          NaN     NaN        NaN    NaN     NaN   
50%                     NaN          NaN     NaN        NaN    NaN     NaN   
75%                     NaN          NaN     NaN        NaN    NaN     NaN   
max                     NaN          NaN     NaN        NaN    NaN     NaN   

       vehicle_type  registration_year  gearbox       powerps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

Columns with only one value: 'nr_of_pictures' (we will drop this)

Columns with two values and need further investigation: 'seller', 'offer_type', 'ab_test', 'gearbox','unrepaired damage'

Columns that are numeric data stored as text: 'price' and 'odometer' (we will work to remove non-numeric characters, convert the column to a numeric dtype, and rename the column to 'odometer_km')

In [349]:
autos['seller'].value_counts()

privat        49999
gewerblich        1
Name: seller, dtype: int64

In [350]:
autos['seller'].head()

0    privat
1    privat
2    privat
3    privat
4    privat
Name: seller, dtype: object

In [351]:
#we will drop the seller column. Let us check the rest
autos['offer_type'].value_counts()

Angebot    49999
Gesuch         1
Name: offer_type, dtype: int64

In [352]:
autos['ab_test'].value_counts()


test       25756
control    24244
Name: ab_test, dtype: int64

In [353]:
autos['gearbox'].value_counts()


manuell      36993
automatik    10327
Name: gearbox, dtype: int64

In [354]:
autos['unrepaired_damage'].value_counts()

nein    35232
ja       4939
Name: unrepaired_damage, dtype: int64

The columns 'seller' and 'offer_type' are mostly made up of one value only. We will seek to drop these columns.

In [355]:
#we will now convert price and odometer to numeric columns
print(autos['price'].value_counts())

$0         1421
$500        781
$1,500      734
$2,500      643
$1,000      639
           ... 
$3,620        1
$6,195        1
$66,500       1
$13,362       1
$9,130        1
Name: price, Length: 2357, dtype: int64


In [356]:
autos['price'].dtype

dtype('O')

In [357]:
autos['odometer'].dtype

dtype('O')

In [358]:
print(autos['odometer'].value_counts())

150,000km    32424
125,000km     5170
100,000km     2169
90,000km      1757
80,000km      1436
70,000km      1230
60,000km      1164
50,000km      1027
5,000km        967
40,000km       819
30,000km       789
20,000km       784
10,000km       264
Name: odometer, dtype: int64


In [359]:
autos.dtypes

date_crawled          object
name                  object
seller                object
offer_type            object
price                 object
ab_test               object
vehicle_type          object
registration_year      int64
gearbox               object
powerps                int64
model                 object
odometer              object
registration_month     int64
fuel_type             object
brand                 object
unrepaired_damage     object
ad_created            object
nr_of_pictures         int64
postal_code            int64
last_seen             object
dtype: object

In [360]:
autos['price'] = autos['price'].str.replace(',','')
autos['price'] = autos['price'].str.replace('$','')
autos['price'] = autos['price'].astype(int)

In [361]:
autos['odometer'] = autos['odometer'].str.replace(',','')
autos['odometer'] = autos['odometer'].str.replace('km','')
autos['odometer'] = autos['odometer'].astype(int)
autos.rename({'odometer':'odometer_km'}, axis = 1, inplace = True)

In [362]:
autos.dtypes

date_crawled          object
name                  object
seller                object
offer_type            object
price                  int64
ab_test               object
vehicle_type          object
registration_year      int64
gearbox               object
powerps                int64
model                 object
odometer_km            int64
registration_month     int64
fuel_type             object
brand                 object
unrepaired_damage     object
ad_created            object
nr_of_pictures         int64
postal_code            int64
last_seen             object
dtype: object

Now that we have converted 'price' and 'odometer' columns to numeric dtypes, we will continue to analyze these columns for any values that don't look correct. We will do this using a few different techniques as follows.

In [363]:
autos['odometer_km'].unique().shape

(13,)

In [364]:
autos['odometer_km'].describe()

count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

In [365]:
autos['odometer_km'].value_counts().head(13)

150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
5000        967
40000       819
30000       789
20000       784
10000       264
Name: odometer_km, dtype: int64

In [366]:
autos['price'].unique().shape

(2357,)

In [367]:
autos['price'].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

In [368]:
selection = autos['price'].value_counts(dropna = False).sort_index(ascending = False).head(10)
print(selection)

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
Name: price, dtype: int64


In [369]:
autos['price'].mean()

9840.04376

In [370]:
selection = autos[autos['price'].between(500,500000)]['price']
print(selection)
selection.min()

0         5000
1         8500
2         8990
3         4350
4         1350
         ...  
49995    24900
49996     1980
49997    13200
49998    22900
49999     1250
Name: price, Length: 45097, dtype: int64


500

In [371]:
autos['price'] = selection

In [372]:
autos['price'].describe()

count     45097.000000
mean       6320.659600
std        9261.841444
min         500.000000
25%        1500.000000
50%        3500.000000
75%        7900.000000
max      350000.000000
Name: price, dtype: float64

There appear to be no outliers in the 'odometer_km' column. However, the price column has values as low as 1 and as high as 99999999. Due to this, we will remove vales less than 500 and over 500000, as it is unlikely that an auto was sold below or over this price, respectively.

We will now analyze five columns that contain date information. Three of these columns are identified as strings and two are represented as numeric values. We will use a series of methods to understand the date ranges for these columns. 

In [373]:
date_crawled = (autos['date_crawled']
                .str.split()
                .str[0]
               )

In [374]:
date_crawled.value_counts().sort_index()

2016-03-05    1269
2016-03-06     697
2016-03-07    1798
2016-03-08    1665
2016-03-09    1661
2016-03-10    1606
2016-03-11    1624
2016-03-12    1839
2016-03-13     778
2016-03-14    1831
2016-03-15    1699
2016-03-16    1475
2016-03-17    1576
2016-03-18     653
2016-03-19    1745
2016-03-20    1891
2016-03-21    1876
2016-03-22    1647
2016-03-23    1619
2016-03-24    1455
2016-03-25    1587
2016-03-26    1624
2016-03-27    1552
2016-03-28    1742
2016-03-29    1709
2016-03-30    1681
2016-03-31    1596
2016-04-01    1690
2016-04-02    1770
2016-04-03    1934
2016-04-04    1826
2016-04-05     655
2016-04-06     159
2016-04-07      71
Name: date_crawled, dtype: int64

In [375]:
date_crawled.value_counts(normalize = True, dropna = False).sort_index()

2016-03-05    0.02538
2016-03-06    0.01394
2016-03-07    0.03596
2016-03-08    0.03330
2016-03-09    0.03322
2016-03-10    0.03212
2016-03-11    0.03248
2016-03-12    0.03678
2016-03-13    0.01556
2016-03-14    0.03662
2016-03-15    0.03398
2016-03-16    0.02950
2016-03-17    0.03152
2016-03-18    0.01306
2016-03-19    0.03490
2016-03-20    0.03782
2016-03-21    0.03752
2016-03-22    0.03294
2016-03-23    0.03238
2016-03-24    0.02910
2016-03-25    0.03174
2016-03-26    0.03248
2016-03-27    0.03104
2016-03-28    0.03484
2016-03-29    0.03418
2016-03-30    0.03362
2016-03-31    0.03192
2016-04-01    0.03380
2016-04-02    0.03540
2016-04-03    0.03868
2016-04-04    0.03652
2016-04-05    0.01310
2016-04-06    0.00318
2016-04-07    0.00142
Name: date_crawled, dtype: float64

In [376]:
date_crawled.value_counts(normalize = True, dropna = False).sort_values()

2016-04-07    0.00142
2016-04-06    0.00318
2016-03-18    0.01306
2016-04-05    0.01310
2016-03-06    0.01394
2016-03-13    0.01556
2016-03-05    0.02538
2016-03-24    0.02910
2016-03-16    0.02950
2016-03-27    0.03104
2016-03-17    0.03152
2016-03-25    0.03174
2016-03-31    0.03192
2016-03-10    0.03212
2016-03-23    0.03238
2016-03-26    0.03248
2016-03-11    0.03248
2016-03-22    0.03294
2016-03-09    0.03322
2016-03-08    0.03330
2016-03-30    0.03362
2016-04-01    0.03380
2016-03-15    0.03398
2016-03-29    0.03418
2016-03-28    0.03484
2016-03-19    0.03490
2016-04-02    0.03540
2016-03-07    0.03596
2016-04-04    0.03652
2016-03-14    0.03662
2016-03-12    0.03678
2016-03-21    0.03752
2016-03-20    0.03782
2016-04-03    0.03868
Name: date_crawled, dtype: float64

The date_crawled column contains dates ranging from March 03, 2016 to April 07, 2016. The date with the highest frequency is April 04, 2016, with a percentage of 3.868%.

In [377]:
ad_created = (autos['ad_created']
                .str.split()
                .str[0]
               )

In [378]:
ad_created.value_counts(normalize = True, dropna = False).sort_index()

2015-06-11    0.00002
2015-08-10    0.00002
2015-09-09    0.00002
2015-11-10    0.00002
2015-12-05    0.00002
               ...   
2016-04-03    0.03892
2016-04-04    0.03688
2016-04-05    0.01184
2016-04-06    0.00326
2016-04-07    0.00128
Name: ad_created, Length: 76, dtype: float64

In [379]:
ad_created.value_counts(normalize = True, dropna = False).sort_values()

2016-02-08    0.00002
2016-02-01    0.00002
2016-01-29    0.00002
2016-02-11    0.00002
2016-02-22    0.00002
               ...   
2016-03-12    0.03662
2016-04-04    0.03688
2016-03-21    0.03772
2016-03-20    0.03786
2016-04-03    0.03892
Name: ad_created, Length: 76, dtype: float64

In the 'ad_created' column, we can see that the dates range from June 11, 2015 to April 07, 2016. Once more, April 03, 2016 has the highest frequency, with a percentage of 3.892%. 

In [380]:
last_seen = (autos['last_seen']
                .str.split()
                .str[0]
               )

In [381]:
last_seen.value_counts(normalize = True, dropna = False).sort_index()

2016-03-05    0.00108
2016-03-06    0.00442
2016-03-07    0.00536
2016-03-08    0.00760
2016-03-09    0.00986
2016-03-10    0.01076
2016-03-11    0.01252
2016-03-12    0.02382
2016-03-13    0.00898
2016-03-14    0.01280
2016-03-15    0.01588
2016-03-16    0.01644
2016-03-17    0.02792
2016-03-18    0.00742
2016-03-19    0.01574
2016-03-20    0.02070
2016-03-21    0.02074
2016-03-22    0.02158
2016-03-23    0.01858
2016-03-24    0.01956
2016-03-25    0.01920
2016-03-26    0.01696
2016-03-27    0.01602
2016-03-28    0.02086
2016-03-29    0.02234
2016-03-30    0.02484
2016-03-31    0.02384
2016-04-01    0.02310
2016-04-02    0.02490
2016-04-03    0.02536
2016-04-04    0.02462
2016-04-05    0.12428
2016-04-06    0.22100
2016-04-07    0.13092
Name: last_seen, dtype: float64

In [382]:
last_seen.value_counts(normalize = True, dropna = False).sort_values()

2016-03-05    0.00108
2016-03-06    0.00442
2016-03-07    0.00536
2016-03-18    0.00742
2016-03-08    0.00760
2016-03-13    0.00898
2016-03-09    0.00986
2016-03-10    0.01076
2016-03-11    0.01252
2016-03-14    0.01280
2016-03-19    0.01574
2016-03-15    0.01588
2016-03-27    0.01602
2016-03-16    0.01644
2016-03-26    0.01696
2016-03-23    0.01858
2016-03-25    0.01920
2016-03-24    0.01956
2016-03-20    0.02070
2016-03-21    0.02074
2016-03-28    0.02086
2016-03-22    0.02158
2016-03-29    0.02234
2016-04-01    0.02310
2016-03-12    0.02382
2016-03-31    0.02384
2016-04-04    0.02462
2016-03-30    0.02484
2016-04-02    0.02490
2016-04-03    0.02536
2016-03-17    0.02792
2016-04-05    0.12428
2016-04-07    0.13092
2016-04-06    0.22100
Name: last_seen, dtype: float64

In the 'last_seen' column, we can see that the dates range from March 05, 2016 to April 06, 2016. Additionally, the date of April 06, 2016 has the highest frequency, with a percentage of 22.10%.

In [383]:
autos['registration_year'].describe()

count    50000.000000
mean      2005.073280
std        105.712813
min       1000.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

When we observe the column 'registration_year', we can see that there are outliers. The min value is 1000 and the max value is 9999. This is erroneous as there were no cars in the year 1000 and we are far from the year 9999.

In [384]:
autos['registration_year'].unique().shape

(97,)

In [385]:
autos['registration_year'].value_counts()

2000    3354
2005    3015
1999    3000
2004    2737
2003    2727
        ... 
1931       1
1929       1
1001       1
9996       1
1952       1
Name: registration_year, Length: 97, dtype: int64

In [386]:
autos['registration_year'].value_counts().sort_index().head(20)

1000    1
1001    1
1111    1
1500    1
1800    2
1910    9
1927    1
1929    1
1931    1
1934    2
1937    4
1938    1
1939    1
1941    2
1943    1
1948    1
1950    3
1951    2
1952    1
1953    1
Name: registration_year, dtype: int64

In [387]:
autos['registration_year'].value_counts().sort_values(ascending = False)

2000    3354
2005    3015
1999    3000
2004    2737
2003    2727
        ... 
1111       1
1953       1
4800       1
4100       1
1952       1
Name: registration_year, Length: 97, dtype: int64

Any vehicle with a registration year above 2016 is inaccurate because a car can't first be registered after a listing was seen. Furthermore, a vehicle can't be registered before 1910, since that is the earliest year value before jumping to 1800, when cars did not exist yet. We will remove all values below 1910 and above 2016.

In [388]:
autos['registration_year'] = autos[autos['registration_year'].between(1910,2016)]['registration_year']

In [389]:
autos['registration_year'].value_counts(normalize = True).sort_values()

1927.0    0.000021
1952.0    0.000021
1953.0    0.000021
1948.0    0.000021
1929.0    0.000021
            ...   
2003.0    0.056779
2004.0    0.056988
1999.0    0.062464
2005.0    0.062776
2000.0    0.069834
Name: registration_year, Length: 78, dtype: float64

We can see that the majority of autos were registered in the late 1990s or early 2000s. Specifically, the year 2000 has te highest frequency.

We will now explore the 'brand' column. Specifically, we will be aggregating over certain values in this column. Let's choose the top 20 car brands in terms of frequency for this exercise.

In [392]:
unique_brand = autos['brand'].unique()
print(unique_brand)

['peugeot' 'bmw' 'volkswagen' 'smart' 'ford' 'chrysler' 'seat' 'renault'
 'mercedes_benz' 'audi' 'sonstige_autos' 'opel' 'mazda' 'porsche' 'mini'
 'toyota' 'dacia' 'nissan' 'jeep' 'saab' 'volvo' 'mitsubishi' 'jaguar'
 'fiat' 'skoda' 'subaru' 'kia' 'citroen' 'chevrolet' 'hyundai' 'honda'
 'daewoo' 'suzuki' 'trabant' 'land_rover' 'alfa_romeo' 'lada' 'rover'
 'daihatsu' 'lancia']


In [420]:
autos['brand'].value_counts(normalize = True).sort_values(ascending = False).head(20)

volkswagen        0.21374
opel              0.10922
bmw               0.10858
mercedes_benz     0.09468
audi              0.08566
ford              0.06958
renault           0.04808
peugeot           0.02912
fiat              0.02616
seat              0.01882
skoda             0.01572
mazda             0.01514
nissan            0.01508
citroen           0.01402
smart             0.01402
toyota            0.01234
sonstige_autos    0.01092
hyundai           0.00976
volvo             0.00914
mini              0.00848
Name: brand, dtype: float64

In [421]:
autos['brand'].value_counts(normalize = True).sort_values(ascending = False).head(20).index

Index(['volkswagen', 'opel', 'bmw', 'mercedes_benz', 'audi', 'ford', 'renault',
       'peugeot', 'fiat', 'seat', 'skoda', 'mazda', 'nissan', 'citroen',
       'smart', 'toyota', 'sonstige_autos', 'hyundai', 'volvo', 'mini'],
      dtype='object')

In [403]:
brand_top20 = ['volkswagen', 'opel', 'bmw', 'mercedes_benz', 'audi', 'ford', 'renault',
       'peugeot', 'fiat', 'seat', 'skoda', 'mazda', 'nissan', 'citroen',
       'smart', 'toyota', 'sonstige_autos', 'hyundai', 'volvo', 'mini']

In [404]:
print(brand_top20)

['volkswagen', 'opel', 'bmw', 'mercedes_benz', 'audi', 'ford', 'renault', 'peugeot', 'fiat', 'seat', 'skoda', 'mazda', 'nissan', 'citroen', 'smart', 'toyota', 'sonstige_autos', 'hyundai', 'volvo', 'mini']


In [413]:
brand_means = {}
for i in brand_top20:
    selected_rows = autos[autos['brand'] == i]
    mean = selected_rows['price'].mean() #can also do selected_rows['price'].sum() / selected_rows['price'].count()
    brand_means[i] = mean

In [419]:
brand_means

{'volkswagen': 5701.241955574009,
 'opel': 3348.936054124836,
 'bmw': 8514.81103286385,
 'mercedes_benz': 8670.394058540847,
 'audi': 9484.499876390606,
 'ford': 4267.123376623376,
 'renault': 2760.8114143920598,
 'peugeot': 3329.3254211332314,
 'fiat': 3211.992606284658,
 'seat': 4709.527446300716,
 'skoda': 6539.7107001321,
 'mazda': 4391.736377025037,
 'nissan': 5077.141592920354,
 'citroen': 3994.679626749611,
 'smart': 3552.050946142649,
 'toyota': 5244.499165275459,
 'sonstige_autos': 13200.66974595843,
 'hyundai': 5695.464757709251,
 'volvo': 5110.575539568345,
 'mini': 10616.96626506024}

After creating a new dictionary with the mean price for the top 20 car brands, we can see that Sonstige Autos has the highest mean price at 13,200. Meanwhile, Renault has the lowest mean price as 2,760. We will repeat the same process to calculate mean mileage for the top 20 brands.

In [433]:
brand_mileage = {}
for i in brand_top20:
    selected_rows = autos[autos['brand'] == i]
    mean = selected_rows['odometer_km'].mean() 
    brand_mileage[i] = mean

In [440]:
brand_mileage

{'volkswagen': 128955.27276129878,
 'opel': 129298.66324848929,
 'bmw': 132521.64302818198,
 'mercedes_benz': 130886.14279678918,
 'audi': 129643.9411627364,
 'ford': 124131.93446392642,
 'renault': 128223.79367720465,
 'peugeot': 127352.33516483517,
 'fiat': 117037.4617737003,
 'seat': 122061.63655685441,
 'skoda': 110947.83715012722,
 'mazda': 125132.10039630119,
 'nissan': 118978.7798408488,
 'citroen': 119764.62196861627,
 'smart': 100756.06276747503,
 'toyota': 115988.65478119935,
 'sonstige_autos': 87188.6446886447,
 'hyundai': 106782.7868852459,
 'volvo': 138632.3851203501,
 'mini': 89375.0}

Now, we will use the Series contructor to convert the brand_means and brand_mileage dictionary into a series that separates the key from the value. The key becomes the index in the series object. Then, we will use the Dataframe constructor to create a single-column dataframe from this series object. We will use the columns parameter to specify the column name as "mean_price" and "mean_mileage"

In [442]:
bm_series = pd.Series(brand_means)
print(bm_series)

volkswagen         5701.241956
opel               3348.936054
bmw                8514.811033
mercedes_benz      8670.394059
audi               9484.499876
ford               4267.123377
renault            2760.811414
peugeot            3329.325421
fiat               3211.992606
seat               4709.527446
skoda              6539.710700
mazda              4391.736377
nissan             5077.141593
citroen            3994.679627
smart              3552.050946
toyota             5244.499165
sonstige_autos    13200.669746
hyundai            5695.464758
volvo              5110.575540
mini              10616.966265
dtype: float64


In [443]:
bm_df = pd.DataFrame(bm_series, columns = ['mean_price'])
bm_df

mean_price
volkswagen       5701.241956
opel             3348.936054
bmw              8514.811033
mercedes_benz    8670.394059
audi             9484.499876
ford             4267.123377
renault          2760.811414
peugeot          3329.325421
fiat             3211.992606
seat             4709.527446
skoda            6539.710700
mazda            4391.736377
nissan           5077.141593
citroen          3994.679627
smart            3552.050946
toyota           5244.499165
sonstige_autos  13200.669746
hyundai          5695.464758
volvo            5110.575540
mini            10616.966265

In [444]:
mean_mileage = pd.Series(brand_mileage)
print(mean_mileage)

volkswagen        128955.272761
opel              129298.663248
bmw               132521.643028
mercedes_benz     130886.142797
audi              129643.941163
ford              124131.934464
renault           128223.793677
peugeot           127352.335165
fiat              117037.461774
seat              122061.636557
skoda             110947.837150
mazda             125132.100396
nissan            118978.779841
citroen           119764.621969
smart             100756.062767
toyota            115988.654781
sonstige_autos     87188.644689
hyundai           106782.786885
volvo             138632.385120
mini               89375.000000
dtype: float64


In [445]:
bm_df['mean_mileage'] = mean_mileage
bm_df

mean_price   mean_mileage
volkswagen       5701.241956  128955.272761
opel             3348.936054  129298.663248
bmw              8514.811033  132521.643028
mercedes_benz    8670.394059  130886.142797
audi             9484.499876  129643.941163
ford             4267.123377  124131.934464
renault          2760.811414  128223.793677
peugeot          3329.325421  127352.335165
fiat             3211.992606  117037.461774
seat             4709.527446  122061.636557
skoda            6539.710700  110947.837150
mazda            4391.736377  125132.100396
nissan           5077.141593  118978.779841
citroen          3994.679627  119764.621969
smart            3552.050946  100756.062767
toyota           5244.499165  115988.654781
sonstige_autos  13200.669746   87188.644689
hyundai          5695.464758  106782.786885
volvo            5110.575540  138632.385120
mini            10616.966265   89375.000000

In [447]:
#sort by mean_price
bm_df.sort_values(by = ['mean_mileage'])

mean_price   mean_mileage
sonstige_autos  13200.669746   87188.644689
mini            10616.966265   89375.000000
smart            3552.050946  100756.062767
hyundai          5695.464758  106782.786885
skoda            6539.710700  110947.837150
toyota           5244.499165  115988.654781
fiat             3211.992606  117037.461774
nissan           5077.141593  118978.779841
citroen          3994.679627  119764.621969
seat             4709.527446  122061.636557
ford             4267.123377  124131.934464
mazda            4391.736377  125132.100396
peugeot          3329.325421  127352.335165
renault          2760.811414  128223.793677
volkswagen       5701.241956  128955.272761
opel             3348.936054  129298.663248
audi             9484.499876  129643.941163
mercedes_benz    8670.394059  130886.142797
bmw              8514.811033  132521.643028
volvo            5110.575540  138632.385120

Once we sort by mean mileage, we can see that there is an inverse correlation between mean price and mean mileage. We can deduce that as mean mileage goes up, the mean price goes down for the brand.

# Conclusion

In conclusion, we were able to conduct a brief analysis of a dataset of used cars from the classifieds section of the German eBay website. Throughout this project, we were able to clean the data by replacing symbols, removing spaces, renaming columns, and dropping outliers. Once the data was clean, we were able to identify the top 20 auto brands and conduct a brief analysis on their mean price and mean mileage.